vmm/src/serial_manager.rs

```diff
@@ -0,0 +1,235 @@
+            epoll::Event::new(epoll::Events::EPOLLIN, dispatch_index),
+        )?;
+
+        Ok(())
+    }
+
+    fn add_event<T>(&mut self, fd: &T, token: EpollDispatch) -> result::Result<(), io::Error>
+    where
+        T: AsRawFd,
+    {
+        let dispatch_index = token as u64;
+        epoll::ctl(
+            self.epoll_file.as_raw_fd(),
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            fd.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, dispatch_index),
+        )?;
+
+        Ok(())
+    }
+}
+
+impl AsRawFd for EpollContext {
+    fn as_raw_fd(&self) -> RawFd {
+        self.epoll_file.as_raw_fd()
+    }
+}
+
+pub struct SerialManager {
+    #[cfg(target_arch = "x86_64")]
+    serial: Arc<Mutex<Serial>>,
+    #[cfg(target_arch = "aarch64")]
+    serial: Arc<Mutex<Pl011>>,
+    epoll: EpollContext,
+    pty: Option<File>,
+}
+
+impl SerialManager {
+    #[cfg(target_arch = "x86_64")]
+    pub fn new(serial: Arc<Mutex<Serial>>, pty: Option<File>, serial_tty: bool) -> Result<Self> {
+        let mut epoll = EpollContext::new().map_err(Error::Epoll)?;
+        let pty = if pty.is_some() {
+            let pty_clone = pty.unwrap().try_clone().map_err(Error::FileClone)?;
+            epoll
+                .add_event(&pty_clone, EpollDispatch::SerialPty)
+                .map_err(Error::Epoll)?;
+            Some(pty_clone)
+        } else {
+            None
+        };
+        if serial_tty && unsafe { libc::isatty(libc::STDIN_FILENO as i32) } != 0 {
+            epoll.add_stdin().map_err(Error::Epoll)?;
+        }
+
+        Ok(SerialManager { serial, epoll, pty })
+    }
+
+    #[cfg(target_arch = "aarch64")]
+    pub fn new(serial: Arc<Mutex<Pl011>>, pty: Option<File>, serial_tty: bool) -> Result<Self> {
+        let mut epoll = EpollContext::new().map_err(Error::Epoll)?;
+        let pty = if pty.is_some() {
+            let pty_clone = pty.unwrap().try_clone().map_err(Error::FileClone)?;
+            epoll
+                .add_event(&pty_clone, EpollDispatch::SerialPty)
+                .map_err(Error::Epoll)?;
+            Some(pty_clone)
+        } else {
+            None
+        };
+        if serial_tty && unsafe { libc::isatty(libc::STDIN_FILENO as i32) } != 0 {
+            epoll.add_stdin().map_err(Error::Epoll)?;
+        }
+
+        Ok(SerialManager { serial, epoll, pty })
+    }
+
+    fn queue_input_bytes_serial(&self, out: &[u8]) -> Result<()> {
+        self.serial
+            .as_ref()
+            .lock()
+            .unwrap()
+            .queue_input_bytes(out)
+            .map_err(Error::QueueInput)?;
+
+        Ok(())
+    }
+
+    fn handle_pty(&mut self) -> Result<()> {
+        let mut out = [0u8; 64];
+        if let Some(pty) = &mut self.pty {
+            let count = pty.read(&mut out).map_err(Error::Pty)?;
+            self.queue_input_bytes_serial(&out[..count])?;
+        }
+
+        Ok(())
+    }
+
+    fn handle_stdin(&self) -> Result<()> {
```

Comment:
You only need one function to read from the active `File` either PTY, TTY (stdin) or real file on disk.